src/main/java/picard/sam/markduplicates/UmiGraph.java

```diff
@@ -0,0 +1,210 @@
+        int root = p;
+        while (root != duplicateSet[root])
+            root = duplicateSet[root];
+        while (p != root) {
+            int newp = duplicateSet[p];
+            duplicateSet[p] = root;
+            p = newp;
+        }
+        return root;
+    }
+
+    public void union(int p, int q) {
+        int rootP = find(p);
+        int rootQ = find(q);
+        if (rootP == rootQ) return;
+        duplicateSet[rootP] = rootQ;
+        nDuplicateSets--;
+    }
+
+    List<DuplicateSet> joinUmisIntoDuplicateSets(List<SAMRecord> records, String umiTag, String inferredUmiTag) {
+        // Iterate over diagonal
+        for (int i = 0; i < nUmis; i++) {
+            for (int j = i + 1; j < nUmis; j++) {
+                if (getEditDistance(umi[i], umi[j]) <= editDistanceToJoin) {
+                    union(i, j);
+                }
+            }
+        }
+
+        for (int i = 0; i < nUmis; i++) {
+            duplicateSet[i] = find(i);
+        }
+
+        Map<Integer, List<SAMRecord>> duplicateSets = new HashMap<>();
+
+        // Assign UMIs to duplicateSets
+        Map<String, Integer> duplicateSetsFromUmis = getDuplicateSetsFromUmis();
+        for(SAMRecord rec : records) {
+
+            String umi = rec.getStringAttribute(umiTag);
+            Integer duplicateSetIndex = duplicateSetsFromUmis.get(umi);
+
+            if(duplicateSets.containsKey(duplicateSetIndex)) {
+                duplicateSets.get(duplicateSetIndex).add(rec);
+            }
+            else {
+                List<SAMRecord> n = new ArrayList();
+                n.add(rec);
+                duplicateSets.put(duplicateSetIndex, n);
+            }
+        }
+
+
+        List<DuplicateSet> duplicateSetList = new ArrayList();
+        for (Map.Entry<Integer, List<SAMRecord>> entry : duplicateSets.entrySet()) {
+            DuplicateSet ds = new DuplicateSet();
+            List<SAMRecord> recordList = entry.getValue();
+
+            // Add records to the DuplicateSet
+            for(SAMRecord rec : recordList) {
+                ds.add(rec);
+            }
+
+            // For a particular duplicate set, identify the most common umi
+            // and use this an an iferred umi
+            Long maxCount = new Long(0);
+            String inferredUmi = null;
+            for(SAMRecord rec : recordList) {
+                String umi = rec.getStringAttribute(umiTag);
+                //inferredUmi = umi;
+                if(umiCounts.get(umi) > maxCount) {
+                   maxCount = umiCounts.get(umi);
+                   inferredUmi = umi;
+                }
+            }
+
+            // Set the records to contain the inferred Umi
+            for(SAMRecord rec : recordList) {
+                rec.setAttribute(inferredUmiTag, inferredUmi);
+            }
+
+            duplicateSetList.add(ds);
+        }
+
+        return duplicateSetList;
+    }
+
+    private int getEditDistance(final String s1, final String s2) {
```

Comment:
you might want to profile the code to see if comparing arrays of bytes would be faster.